File: src/twg/ui/widgets/help.py
import importlib.metadata
from textual.app import ComposeResult
from textual.screen import ModalScreen
from textual.widgets import Label
from textual.containers import Container, Grid

class HelpModal(ModalScreen[None]):
    """Modal screen giving help/keybindings."""

    LOGO = r"""
████████╗██╗    ██╗██╗ ██████╗ 
╚══██╔══╝██║    ██║██║██╔════╝ 
   ██║   ██║ █╗ ██║██║██║  ███╗
   ██║   ██║███╗██║██║██║   ██║
   ██║   ╚███╔███╔╝██║╚██████╔╝
   ╚═╝    ╚══╝╚══╝ ╚═╝ ╚═════╝ 
"""

    CSS = """
    HelpModal {
        align: center middle;
    }

    #help-container {
        width: 70;
        height: auto;
        background: $surface;
        border: tall $primary; 
        padding: 1 2;
        align: center middle;
    }

    #help-logo {
        width: 100%;
        text-align: center;
        color: $warning;
        margin-top: 1;
        margin-bottom: 1;
    }

    .help-tagline {
        width: 100%;
        text-align: center;
        color: $success;
        text-style: bold;
        margin-bottom: 1;
    }

    .help-version {
        width: 100%;
        text-align: center;
        color: $text-muted;
        margin-bottom: 2;
    }

    #help-link {
        width: 100%;
        text-align: center;
        color: $primary;
        text-style: underline;
        margin-bottom: 1;
    }

    #coffee-link {
        width: 100%;
        text-align: center;
        color: $accent;
        margin-bottom: 2;
    }

    #help-title {
        width: 100%;
        text-align: center;
        text-style: bold underline;
        color: $text;
        margin-bottom: 1;
        margin-top: 1;
        background: $surface;
    }

    #help-grid {
        layout: grid;
        grid-size: 2;
        grid-columns: 1fr 1fr;
        grid-gutter: 0 2;
        width: 100%;
        height: auto;
        margin-bottom: 2;
        align: center middle;
    }

    .key {
        text-align: left;
        text-style: bold;
        color: $accent;
    }

    .desc {
        text-align: right;
        color: $text;
    }

    #help-exit {
        width: 100%;
        text-align: center;
        color: $text-muted;
        margin-bottom: 1;
    }
    """

    KEYBINDINGS = [
        ("Arrows", "Navigate tree"),
        ("/", "Search "),
        ("n / N", "Next / Prev match"),
        (":", "Jump to path"),
        ("c", "Copy path"),
        ("y", "Copy source"),
        ("t", "Toggle theme"),
        ("q", "Quit"),
    ]

    def compose(self) -> ComposeResult:
        try:
            version = importlib.metadata.version("twg")
        except importlib.metadata.PackageNotFoundError:
            version = "unknown"

        with Container(id="help-container"):
            # Branding Section
            yield Label(self.LOGO, id="help-logo")
            yield Label("Inspect. Navigate. Understand.", classes="help-tagline")
            yield Label(f"v{version}", classes="help-version")
            yield Label(f"v{version}", classes="help-version")
            yield Label("https://twig.wtf", id="help-link")
            yield Label("☕ buymeacoffee.com/workdone0", classes="help-tagline", id="coffee-link")
            
            # Keybindings Section
            yield Label("Keyboard Shortcuts", id="help-title")
            
            with Grid(id="help-grid"):
                for key, desc in self.KEYBINDINGS:
                    yield Label(key, classes="key")
                    yield Label(desc, classes="desc")

            yield Label("(Press Esc to close)", id="help-exit")

    def key_escape(self) -> None:
        self.dismiss(None)
    

File: src/tui/widgets/help.rs
//! Centered help modal.

use ratatui::layout::{Alignment, Constraint, Direction, Layout, Rect};
use ratatui::style::{Modifier, Style};
use ratatui::text::{Line, Span};
use ratatui::widgets::{Block, Borders, Clear, Paragraph, Wrap};
use ratatui::Frame;

use crate::tui::theme::Theme;

pub const LOGO: &str = r"
████████╗██╗    ██╗██╗ ██████╗
╚══██╔══╝██║    ██║██║██╔════╝
   ██║   ██║ █╗ ██║██║██║  ███╗
   ██║   ██║███╗██║██║██║   ██║
   ██║   ╚███╔███╔╝██║╚██████╔╝
   ╚═╝    ╚══╝╚══╝ ╚═╝ ╚═════╝
";

const KEYBINDINGS: &[(&str, &str)] = &[
    ("Arrows", "Navigate tree"),
    ("/", "Search"),
    ("n / N", "Next / Prev match"),
    (":", "Jump to path"),
    ("c", "Copy path"),
    ("y", "Copy source"),
    ("t", "Toggle theme"),
    ("q", "Quit"),
];

pub fn render(f: &mut Frame, area: Rect, theme: &Theme, version: &str) {
    let modal_w = 60u16.min(area.width.saturating_sub(4));
    let modal_h = 22u16.min(area.height.saturating_sub(2));
    let modal = Rect {
        x: area.x + (area.width.saturating_sub(modal_w)) / 2,
        y: area.y + (area.height.saturating_sub(modal_h)) / 2,
        width: modal_w,
        height: modal_h,
    };

    f.render_widget(Clear, modal);
    let block = Block::default()
        .title(" Help ")
        .borders(Borders::ALL)
        .border_style(Style::default().fg(theme.primary));
    let inner = block.inner(modal);
    f.render_widget(block, modal);

    let rows = Layout::default()
        .direction(Direction::Vertical)
        .constraints([
            Constraint::Length(LOGO.lines().count() as u16),
            Constraint::Length(1),
            Constraint::Length(1),
            Constraint::Length(1),
            Constraint::Length(1),
            Constraint::Length(1),
            Constraint::Length(1),
            Constraint::Length(1),
            Constraint::Min(1),
        ])
        .split(inner);

    f.render_widget(
        Paragraph::new(LOGO)
            .alignment(Alignment::Center)
            .style(Style::default().fg(theme.warning)),
        rows[0],
    );
    f.render_widget(
        Paragraph::new(Line::from("Inspect. Navigate. Understand."))
            .alignment(Alignment::Center)
            .style(Style::default().fg(theme.success).add_modifier(Modifier::BOLD)),
        rows[2],
    );
    f.render_widget(
        Paragraph::new(Line::from(format!("v{version}")))
            .alignment(Alignment::Center)
            .style(Style::default().add_modifier(Modifier::DIM)),
        rows[3],
    );
    f.render_widget(
        Paragraph::new(Line::from("https://twig.wtf"))
            .alignment(Alignment::Center)
            .style(Style::default().fg(theme.primary))
            .wrap(Wrap { trim: false }),
        rows[4],
    );
    f.render_widget(
        Paragraph::new(Line::from(Span::styled(
            "Keyboard Shortcuts",
            Style::default().add_modifier(Modifier::BOLD | Modifier::UNDERLINED),
        )))
        .alignment(Alignment::Center),
        rows[5],
    );

    let mut key_lines: Vec<Line> = Vec::new();
    for (key, desc) in KEYBINDINGS {
        key_lines.push(Line::from(vec![
            Span::styled(format!("  {key:<10}"), Style::default().fg(theme.accent)),
            Span::raw("  "),
            Span::styled(*desc, Style::default().fg(theme.fg)),
        ]));
    }
    f.render_widget(Paragraph::new(key_lines), rows[6]);

    f.render_widget(
        Paragraph::new(Line::from("(Press Esc to close)"))
            .alignment(Alignment::Center)
            .style(Style::default().add_modifier(Modifier::DIM)),
        rows[7],
    );
}